Source/WebGPU/WGSL/ConstantFunctions.h

```diff
@@ -31,21 +31,68 @@
 
 namespace WGSL {
 
-static ConstantValue constantPow(AST::CallExpression&, const FixedVector<ConstantValue>& arguments)
+template<typename Functor, typename... Arguments>
+static ConstantValue scalarOrVector(const Functor& functor, Arguments&&... unpackedArguments)
 {
-    const auto& value = arguments[0];
-    const auto& exponent = arguments[1];
+    unsigned vectorSize = 0;
+    std::initializer_list<ConstantValue> arguments { unpackedArguments... };
+    for (auto argument : arguments) {
+        if (auto* vector = std::get_if<ConstantVector>(&argument)) {
+            vectorSize = vector->elements.size();
+            break;
+        }
+    }
+    if (!vectorSize)
+        return functor(std::forward<Arguments>(unpackedArguments)...);
+
+    constexpr auto argumentCount = sizeof...(Arguments);
+    ConstantVector result(vectorSize);
+    std::array<ConstantValue, argumentCount> scalars;
+    for (unsigned i = 0; i < vectorSize; ++i) {
+        unsigned j = 0;
+        for (const auto& argument : arguments) {
+            if (auto* vector = std::get_if<ConstantVector>(&argument))
+                scalars[j++] = vector->elements[i];
+            else
+                scalars[j++] = argument;
+        }
+        result.elements[i] = std::apply(functor, scalars);
+    }
+    return result;
+
+}
 
-    if (value.isNumber())
+static ConstantValue constantPow(const Type*, const FixedVector<ConstantValue>& arguments)
+{
+    return scalarOrVector([&](auto value, auto exponent) -> ConstantValue {
         return { std::pow(value.toDouble(), exponent.toDouble()) };
+    }, arguments[0], arguments[1]);
+}
 
-    auto& baseVector = std::get<ConstantVector>(value);
-    auto& expVector = std::get<ConstantVector>(exponent);
-    auto size = baseVector.elements.size();
-    ConstantVector result(size);
-    for (unsigned i = 0; i < size; ++i)
-        result.elements[i] = { std::pow(baseVector.elements[i].toDouble(), expVector.elements[i].toDouble()) };
-    return { result };
+static ConstantValue constantMinus(const Type*, const FixedVector<ConstantValue>& arguments)
+{
+    const auto& unaryMinus = [&]() -> ConstantValue {
+        return scalarOrVector([&](const auto& value) -> ConstantValue {
+            if (value.isInt())
+                return -value.toInt();
```

Comment:
Do we have to worry about INT_MIN here? Since we only parse positive integer literals we'd have to have something like `-(-0x7FFFFFFF - 1)`

Reply:
That's a good question... I'm not sure what the spec says about it. I also have to think about how we'd handle

```const x = -INT_MIN``` vs ```const x: i32 = INT_MIN; const y = -x;```

I might leave it for a later PR though, as we need this one to fix the WebGPU sample